core/src/main/java/org/infinispan/functional/impl/ReadWriteMapImpl.java

```diff
@@ -57,7 +58,8 @@ private ReadWriteMapImpl(Params params, FunctionalMapImpl<K, V> functionalMap) {
    @Override
    public <R> CompletableFuture<R> eval(K key, V value, BiFunction<V, ReadWriteEntryView<K, V>, R> f) {
       log.tracef("Invoked eval(k=%s, v=%s, %s)", key, value, params);
-      ReadWriteKeyValueCommand cmd = fmap.commandsFactory.buildReadWriteKeyValueCommand(key, value, f, params);
+      Object keyEncoded = keyDataConversion.toStorage(key);
+      ReadWriteKeyValueCommand cmd = fmap.commandsFactory.buildReadWriteKeyValueCommand(keyEncoded, value, (BiFunction) f, params, keyDataConversion, valueDataConversion);
```

Comment:
On IRC we've discussed that it'll be better to encode the value as well (and decode it before applying the function ofc), because the user type here might be not serializable.